packages/crypto/src/proof.ts

```diff
@@ -0,0 +1,254 @@
+  const payloadDigest = sha256(payloadBytes);
+
+  // Verify ECDSA signature
+  const isValid = p256.verify(signatureBytes, payloadDigest, signingKeyBytes);
+  if (!isValid) {
+    throw new Error("Signature verification failed");
+  }
+}
+
+export async function verifyCertificateChain(
+  cabundle: Uint8Array[],
+  rootCertPem: string,
+  leafCert: Uint8Array,
+  timestampMs: number,
+): Promise<void> {
+  try {
+    // Check root and assert fingerprint
+    const rootX509 = new x509.X509Certificate(rootCertPem);
+    const rootDer = Buffer.from(new Uint8Array(rootX509.rawData));
+    const rootSha = createHash("sha256")
+      .update(rootDer)
+      .digest("hex")
+      .toUpperCase();
+    if (rootSha !== AWS_ROOT_CERT_SHA256) {
+      throw new Error(
+        `Pinned AWS root fingerprint mismatch: expected=${AWS_ROOT_CERT_SHA256} actual=${rootSha}`,
+      );
+    }
+
+    // Bundle starts with root certificate. We're replacing the root with our hardcoded known certificate, so remove first element
+    const bundleWithoutRoot = cabundle.slice(1);
+    const intermediatesX509 = bundleWithoutRoot.map((c) => {
+      if (!c) throw new Error("Invalid certificate data in cabundle");
+      return new x509.X509Certificate(c);
+    });
+    const leaf = new x509.X509Certificate(leafCert);
+
+    // Build path leaf → intermediates → root, with our hardcoded known root certificate
+    const builder = new x509.X509ChainBuilder({
+      certificates: [rootX509, ...intermediatesX509],
+    });
+    const chain = await builder.build(leaf);
+    if (chain.length !== intermediatesX509.length + 2)
+      throw new Error(
+        `Incorrect number of certs in X509 Chain. Expected ${intermediatesX509.length + 2}, got ${chain.length}`,
+      );
+
+    // It appears `builder.build` does signature-only verification of the chain, but there's no commitement to that in the
+    // public api, so we loop through the chain and do (maybe redundant) signature verification
+    const appProofDate = new Date(timestampMs);
+    for (let i = 0; i < chain.length; i++) {
+      const cert = chain[i];
+      if (!cert) throw new Error("Invalid certificate in chain");
+
+      // Verify signature
+      if (i === chain.length - 1) {
+        // is root
+        // Self-signature verification for root certificate
+        const ok = await cert.verify({
+          publicKey: cert.publicKey,
+          date: appProofDate,
+        });
+        if (!ok)
+          throw new Error("Pinned root failed self-signature verification");
+      } else {
+        // Verify signature against issuer
+        const issuer = chain[i + 1];
+        if (!issuer) throw new Error("Issuer can't be null");
+
+        // Attestation docs technically expire after 3 hours, so an app proof generated 3+ hours after an enclave
+        // boots up will fail verification due to certificate expiration. To prevent this failure, we set `signatureOnly: true`.
```

Comment:
Worth mentioning that this is okay to do because enclaves are immutable. Meaning that even if the boot proof is technically invalid because the cert is invalid, the code contained within it cannot change.